graylog2-server/src/main/java/org/graylog/plugins/map/geoip/GeoIpDbFileChangeMonitorService.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.plugins.map.geoip;
+
+import com.google.common.eventbus.EventBus;
+import com.google.common.eventbus.Subscribe;
+import org.graylog.plugins.map.config.DatabaseType;
+import org.graylog.plugins.map.config.DatabaseVendorType;
+import org.graylog.plugins.map.config.GeoIpResolverConfig;
+import org.graylog2.cluster.ClusterConfigChangedEvent;
+import org.graylog2.plugin.cluster.ClusterConfigService;
+import org.graylog2.plugin.utilities.FileInfo;
+import org.graylog2.plugin.validate.ConfigValidationException;
+import org.graylog2.rest.resources.system.GeoIpResolverConfigValidator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import java.nio.file.Paths;
+import java.time.Duration;
+import java.util.EnumMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A service to check whether the ASN and City MaxMind/IPInfo database files have changed, as well as whether the configuration has changed.
+ *
+ * <p>
+ * If the database files have changed, a {@link GeoIpDbFileChangedEvent} is posted, to which {@link org.graylog.plugins.map.geoip.processor.GeoIpProcessor} subscribes and reloads the {@link GeoIpResolverEngine}.
+ * </p>
+ *
+ * <p>
+ * This service also subscribes to {@link ClusterConfigChangedEvent} to update the database files to be monitored, as well as to update the scheduled task ({@link #refreshDatabases()}) which checks for file changes.
+ * </p>
+ */
+@Singleton
+public final class GeoIpDbFileChangeMonitorService {
```

Comment:
The logic in this class looks good and tests successfully, but I think it would be beneficial to convert it to extend AbstractIdleService, which allows it to have `startUp` and `shutDown` lifecycle methods. The main benefit here is that the `registerToEvents()/updateConfiguration()` methods could then be called in the `startUp` method instead of in the constructor. Since the constructors are called in the process on initializing Guice bindings, this isolates the startup logic from the binding init (which prevents any exceptions that might happen in those methods from preventing Guice bindings from completing).

Similar example: https://github.com/Graylog2/graylog2-server/blob/f35df42e165ac570b8b27de3f8eeac85e74ed610/graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/messages/StreamCacheService.java#L41-L67

The `eventBus.unregister(this)` could also be called in `shutDown` to properly unregister it.

I am not certain of this is really required, but the future for the update could also be checked and unscheduled in the `shutDown` method too. I guess this is only really helpful if the future is a long-running operation (or one that could hang). Perhaps this is good to have here, since it involves disk I/O.
https://github.com/Graylog2/graylog-plugin-enterprise/blob/56dd7b641b1c7ba9fbef3ec017a13ab0ac0792e1/enterprise/src/main/java/org/graylog/plugins/license/EnterpriseLicenseChecker.java#L108-L110